File: apps/party-time-backend/src/main/java/com/partytime/configuration/security/TokenAuthentication.java
package com.partytime.configuration.security;

import org.springframework.security.authentication.AbstractAuthenticationToken;

import java.util.Collections;
import java.util.Objects;

public class TokenAuthentication extends AbstractAuthenticationToken {

    private final PartyTimeUserDetails details;

    /**
     * Creates a token with the supplied array of authorities.
     */
    public TokenAuthentication(PartyTimeUserDetails details) {
        super(Collections.emptyList());
        this.details = details;
        setAuthenticated(true);
        setDetails(details);
    }

    @Override
    public String getCredentials() {
        return details.getPassword();
    }

    @Override
    public PartyTimeUserDetails getPrincipal() {
        return details;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        TokenAuthentication that = (TokenAuthentication) o;
        return Objects.equals(getDetails(), that.getDetails());
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), getDetails());
    }
}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/configuration/security/TokenAuthentication.kt
package com.partytime.configuration.security

import org.springframework.security.authentication.AbstractAuthenticationToken
import java.util.Objects

class TokenAuthentication(private val details: PartyTimeUserDetails) : AbstractAuthenticationToken(emptyList()) {
    init {
        isAuthenticated = true
        setDetails(details)
    }

    override fun getCredentials(): String = details.password

    override fun getPrincipal(): PartyTimeUserDetails = details

    override fun equals(other: Any?): Boolean = when {
        this === other -> true
        other == null || javaClass != other.javaClass -> false
        !super.equals(other) -> false
        else -> other is TokenAuthentication && Objects.equals(details, other.details)
    }

    override fun hashCode(): Int = Objects.hash(super.hashCode(), details)
}
